src/Database/LSMTree/Internal/MergingTree.hs

```diff
@@ -0,0 +1,221 @@
+newPendingLevelMerge ::
+     forall m h.
+     (MonadMVar m, MonadMask m, PrimMonad m)
+  => [PreExistingRun m h]
+  -> Maybe (Ref (MergingTree m h))
+  -> m (Ref (MergingTree m h))
+newPendingLevelMerge [] (Just t) = dupRef t
+newPendingLevelMerge prs mmt = do
+    -- There are no interruption points here, and thus provided async exceptions
+    -- are masked then there can be no async exceptions here at all.
+    mergeTreeState <- case (prs, mmt) of
+      ([PreExistingRun r], Nothing) ->
+        CompletedTreeMerge <$> dupRef r
+
+      ([PreExistingMergingRun mr], Nothing) ->
+        OngoingTreeMerge <$> dupRef mr
+
+      _ -> PendingTreeMerge <$>
+            (PendingLevelMerge <$> traverse dupPreExistingRun prs
+                               <*> dupMaybeMergingTree mmt)
+
+    newMergeTree mergeTreeState
+  where
+    dupPreExistingRun (PreExistingRun r) =
+      PreExistingRun <$> dupRef r
+    dupPreExistingRun (PreExistingMergingRun mr) =
+      PreExistingMergingRun <$> dupRef mr
+
+    dupMaybeMergingTree :: Maybe (Ref (MergingTree m h))
+                        -> m (Maybe (Ref (MergingTree m h)))
+    dupMaybeMergingTree Nothing   = return Nothing
+    dupMaybeMergingTree (Just mt) = do
+      isempty <- isStructurallyEmpty mt
+      if isempty
+        then return Nothing
+        else Just <$> dupRef mt
+
+-- | Create a new 'MergingTree' representing the union of one or more merging
+-- trees. This is for unioning the content of multiple tables (represented
+-- themselves as merging trees).
+--
+-- Shape: if all of the input trees are structurally empty, the result will
+-- also be structurally empty. See 'isStructurallyEmpty'.
+--
+-- Resource tracking:
+-- * This allocates a new 'Ref' which the caller is responsible for releasing
+--   eventually.
+-- * The ownership of all input 'Ref's remains with the caller. This action
+--   will create duplicate references, not adopt the given ones.
+--
+-- ASYNC: this should be called with asynchronous exceptions masked because it
+-- allocates\/creates resources.
+newPendingUnionMerge ::
+     (MonadMVar m, MonadMask m, PrimMonad m)
+  => [Ref (MergingTree m h)]
+  -> m (Ref (MergingTree m h))
+newPendingUnionMerge [mt] = dupRef mt
+newPendingUnionMerge mts = do
+    mts' <- mapM dupRef =<< filterM (fmap not . isStructurallyEmpty) mts
+    newMergeTree (PendingTreeMerge (PendingUnionMerge mts'))
+
+-- | Test if a 'MergingTree' is \"obviously\" empty by virtue of its structure.
+-- This is not the same as being empty due to a pending or ongoing merge
+-- happening to produce an empty run.
+--
+isStructurallyEmpty :: MonadMVar m => Ref (MergingTree m h) -> m Bool
+isStructurallyEmpty (DeRef MergingTree {mergeState}) =
+    isEmpty <$> readMVar mergeState
+  where
+    isEmpty (PendingTreeMerge (PendingLevelMerge [] Nothing)) = True
+    isEmpty (PendingTreeMerge (PendingUnionMerge []))         = True
+    isEmpty _                                                 = False
```

Comment:
This function might at some point also be handy for handling the `CompletedTreeMerge r | Run.size r == 0` case. Should that be considered structurally empty? Either way, for now these cases are sufficient.

Reply:
I've added a comment to this effect.